lib/debloater/index.rb

```diff
@@ -0,0 +1,101 @@
+require 'debloater/helpers'
+
+module Debloater
+  class Index
+    include Helpers
+
+    # conn: connection object
+    # method: the function used to obtain index information
+    # name: the index name
+    def initialize(conn, name:, sql:)
+      @conn = conn
+      @name = name
+      @sql = sql
+    end
+
+    attr_reader :name
+
+    def self.each(conn)
+      data = conn.exec_params %{
+        select * from pg_indexes where schemaname = $1
+      }, ['public']
+      data.sort_by { |d| 
+        d['indexname'] 
+      }.each { |d| 
+        yield new(conn, name: d['indexname'], sql: d['indexdef'])
+      }
+    rescue PG::Error => e
+      _fatal e, msg: 'Could not list indexes'
+    end
+
+
+    def pkey?
+      @name =~ /pkey/
+    end
+
+
+    def valid?
+      !_metadata.nil?
+    end
+
+
+    def density
+      @density ||= 0.01 * _metadata['avg_leaf_density'].to_f
+    end
+
+
+    def bloat
+      @bloat ||= size * (1 - density)
+    end
+
+
+    def size
+      @size ||= _metadata['index_size'].to_i / 1024**2
+    end
+
+
+    def debloat!
+      [
+        "DROP INDEX IF EXISTS idx_debloat_new",
+        @sql.
+          sub('CONCURRENTLY', '').
+          sub('CREATE INDEX', 'CREATE INDEX CONCURRENTLY').
+          sub(@name, "idx_debloat_new"),
+        "BEGIN",
+        "ALTER INDEX #{@name} RENAME TO idx_debloat_old",
+        "ALTER INDEX idx_debloat_new RENAME TO #{@name}",
+        "DROP INDEX idx_debloat_old",
+        "COMMIT",
+      ].each do |sql|
+        _log "\t#{sql}"
+        begin
+          @conn.exec sql
+        rescue PG::Error => e
+          _fatal e, msg: "Failure during index debloating, you may need to recover manually."
```

Comment:
This would scare the crap out of me! Any way we could point people to some help about what recovery might be needed so they're not left alone with possibly knackered indexes?